File: solutions/Medium/Binary_Tree_Right_Side_View_0199.py
from typing import List, Optional
from collections import deque

# Definition for a binary tree node.


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        """
        Function: rightSideView
        Description: Returns the values of nodes visible from the right side of a binary tree.

        Parameters:
        - root (Optional[TreeNode]): The root of the binary tree.

        Returns:
        - List[int]: The list of values seen from the right side, from top to bottom.
        """
        if not root:
            return []

        result = []
        queue = deque([root])

        while queue:
            level_size = len(queue)
            for i in range(level_size):
                node = queue.popleft()
                # The rightmost node at each level is added to the result
                if i == level_size - 1:
                    result.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

        return result


# Helper functions for testing
def list_to_binary_tree(lst):
    """Converts a level-order list to a binary tree."""
    if not lst or lst[0] is None:
        return None

    nodes = [TreeNode(val) if val is not None else None for val in lst]

    for i, node in enumerate(nodes):
        if node:
            left_idx = 2 * i + 1
            right_idx = 2 * i + 2
            if left_idx < len(lst):
                node.left = nodes[left_idx]
            if right_idx < len(lst):
                node.right = nodes[right_idx]

    return nodes[0]  # Root node


# Example Test Cases
if __name__ == "__main__":
    solution = Solution()

    # Test case 1: Basic tree with right-side view [1, 3, 4]
    root = list_to_binary_tree([1, 2, 3, None, 5, None, 4])
    print(solution.rightSideView(root))
    # Expected output: [1, 3, 4]

    # Test case 2: Tree with nodes forming deeper right view [1, 3, 4, 5]
    root = list_to_binary_tree([1, 2, 3, 4, None, None, None, 5])
    print(solution.rightSideView(root))
    # Expected output: [1, 3, 4, 5]

    # Test case 3: Simple right-skewed tree [1, 3]
    root = list_to_binary_tree([1, None, 3])
    print(solution.rightSideView(root))
    # Expected output: [1, 3]

    # Test case 4: Empty tree
    root = list_to_binary_tree([])
    print(solution.rightSideView(root))
    # Expected output: []

    # Test case 5: Tree with only left children [1, 2, 3]
    root = list_to_binary_tree([1, 2, None, 3])
    print(solution.rightSideView(root))
    # Expected output: [1, 2, 3]

    # Test case 6: Tree with only right children [1, None, 2, None, 3]
    root = list_to_binary_tree([1, None, 2, None, 3])
    print(solution.rightSideView(root))
    # Expected output: [1, 2, 3]

    print("All test cases passed!")

# Complexity Analysis
# Time Complexity: O(n), where n is the number of nodes in the binary tree.
# We visit each node exactly once, which gives a linear time complexity.
# The maximum number of nodes at any level is n / 2, where n is the number of nodes in the binary tree.
# Hence, the time complexity is O(n).

# Space Complexity: O(n), where n is the number of nodes in the binary tree.
# The maximum space occupied by the queue is the maximum number of nodes at any level in the binary tree.
# In the worst case, the binary tree is a complete binary tree, and the number of nodes at the last level is (n + 1) / 2. Hence, the space complexity is O(n).

File: solutions/Medium/Binary_Tree_Right_Side_View_0199.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * Function: rightSideView
 * Description: This function returns the values of the nodes visible when looking at a binary tree from the right side.
 * - Uses a level-order traversal (BFS) to collect the last node of each level.
 *
 * @param {TreeNode} root - The root of the binary tree.
 * @returns {number[]} - An array containing the rightmost node values from top to bottom.
 */
const rightSideView = function(root) {
    if (!root) return [];

    const result = [];
    const queue = [root];

    while (queue.length > 0) {
        let levelSize = queue.length;

        for (let i = 0; i < levelSize; i++) {
            let node = queue.shift();

            // Capture the last node of each level
            if (i === levelSize - 1) {
                result.push(node.val);
            }

            // Add left and right children to queue
            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
    }

    return result;
};

// Example Test Cases
if (require.main === module) {
    function TreeNode(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }

    // Example Tree 1: [1,2,3,null,5,null,4]
    let root1 = new TreeNode(1);
    root1.left = new TreeNode(2);
    root1.right = new TreeNode(3);
    root1.left.right = new TreeNode(5);
    root1.right.right = new TreeNode(4);

    console.log(rightSideView(root1)); 
    // Expected output: [1, 3, 4]

    // Example Tree 2: [1,2,3,4,null,null,null,5]
    let root2 = new TreeNode(1);
    root2.left = new TreeNode(2);
    root2.right = new TreeNode(3);
    root2.left.left = new TreeNode(4);
    root2.left.left.left = new TreeNode(5);

    console.log(rightSideView(root2));
    // Expected output: [1, 3, 4, 5]

    // Example Tree 3: [1,null,3]
    let root3 = new TreeNode(1);
    root3.right = new TreeNode(3);

    console.log(rightSideView(root3)); 
    // Expected output: [1, 3]

    // Example Tree 4: Empty tree
    console.log(rightSideView(null)); 
    // Expected output: []
}

/*

Complexity Analysis:
- Time Complexity: O(n) - We visit each node once using BFS.
- Space Complexity: O(n) - In the worst case, we store all nodes in the queue (e.g., a complete binary tree).

*/

export default rightSideView;
